cmd/gke-certificates-controller/app/gke_approver.go

```diff
@@ -170,34 +189,171 @@ func hasExactUsages(csr *capi.CertificateSigningRequest, usages []capi.KeyUsage)
 		return false
 	}
 	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
 		return false
 	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+		return false
+	}
+	return true
+}
+
+func isNodeClientCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeCert(opts, csr, x509cr) {
+		return false
+	}
 	if !hasExactUsages(csr, kubeletClientUsages) {
 		return false
 	}
-	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+	return true
+}
+
+func isSelfNodeClientCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeClientCert(opts, csr, x509cr) {
+		return false
+	}
+	if csr.Spec.Username != x509cr.Subject.CommonName {
 		return false
 	}
 	return true
 }
 
-func isSelfNodeClientCert(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
-	if !isNodeClientCert(csr, x509cr) {
+func isNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeCert(opts, csr, x509cr) {
+		return false
+	}
+	if !hasExactUsages(csr, kubeletServerUsages) {
 		return false
 	}
 	if csr.Spec.Username != x509cr.Subject.CommonName {
 		return false
 	}
 	return true
 }
+
+func validateNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	client := oauth2.NewClient(context.Background(), opts.tokenSource)
+
+	if err := validateNodeServerCertInner(client, opts, csr, x509cr); err != nil {
+		glog.Errorf("validating CSR %q: %v", csr.Name, err)
+		return false
+	}
+	return true
+}
+
+// Maximum number of SANs in server CSR.
+const maxServerSANs = 10
+
+// Only check that DNS name and IP in SAN match an existing VM in the project.
+// Username was already checked against CN, so this CSR is coming from
+// authenticated kubelet.
+func validateNodeServerCertInner(client *http.Client, opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) error {
+	numDNS, numIPs := len(x509cr.DNSNames), len(x509cr.IPAddresses)
+	switch {
+	case numDNS != numIPs:
+		return fmt.Errorf("got %d DNS names and %d IPs, want equal numbers", numDNS, numIPs)
+	case numDNS+numIPs == 0:
+		return errors.New("no SAN DNS names and IPs")
+	case numDNS+numIPs > maxServerSANs:
+		return fmt.Errorf("got %d SANs, want at most %d", numDNS, maxServerSANs)
+	case len(x509cr.EmailAddresses) > 0 || len(x509cr.URIs) > 0:
+		return errors.New("only DNS and IP SANs allowed")
+	}
+
+	cs, err := compute.New(client)
+	if err != nil {
+		return fmt.Errorf("creating GCE API client: %v", err)
+	}
+
+	srv := compute.NewInstancesService(cs)
+outer:
+	for _, n := range x509cr.DNSNames {
+	inner:
+		for _, z := range opts.zones {
+			inst, err := srv.Get(opts.projectID, z, n).Do()
```

Comment:
Can you use the common name to find the instance, and find the instance outside the loop?

Reply:
I think we need to loop over zones regardless. The checks above limit how many requests we make.